.github/workflows/test.yml

```diff
@@ -0,0 +1,115 @@
+# Useful GitHub Actions docs:
+#
+# - https://help.github.com/en/actions
+# - https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions
+# - https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow
+# - https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions
+
+name: Continuous Integration
+
+# Only trigger the workflow when pushing to master
+on:
+  - push
+  - pull_request
```

Comment:
So I think in this configuration, this workflow will run on _every_ push, and _every_ opened PR, regardless of destination. Just wanted to check this was intentional?

Reply:
I think this makes sense- it matches what travis does when you enable it on your repo, and means people can run CI locally on their own branches without opening a PR

Reply:
That's fair, just wanted to check :)

Reply:
I think you can leave off -`pull_request` as that is a subset of `push` in this situation

Reply:
The GitHub examples include both:
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-a-list-of-events

Is it related to branches in the same vs different forks?

Reply:
Yeah, from a practical perspective the pull_request event and push event overlap quite a bit as you can imagine.  If you leave both here in this example I think you will burn unnecessary Actions minutes from your free quota.  They are different events but folks will do a push each time they update or open a PR so I think you need only one event in this case.

Reply:
That was my intention, yes. Is there a reason not to do that? Maybe I should leave the `on` blank to use GitHub's own defaults?

Reply:
> That was my intention, yes. Is there a reason not to do that?

why do you want to run parallel tests on every push and on every change to a PR?  I think you only need push if I am understanding correctly?  I could be wrong.  Perhaps start with just `push` and see if that gives you what you are looking for?

Reply:
Running `pull_request` does catch `refs/pull/:prNumber/merge` though which will catch any issues caused by the branch not being up-to-date with master, which I don't think `push` will catch?

Reply:
> why do you want to run parallel tests on every push and on every change to a PR?

Because that's what ~all CI services do by default, right? I just want 'normal' default behavior (i.e. what I've come to expect from Travis, Circle, etc.), which is:

- build head commit of any push to any branch on this repo (won't generally include PRs, which come from forks)
- build merge commit, as @sgibson91 mentions, of any PR on each update / open / etc.

These do tend to duplicate CI runs on PRs from branches from the same repo, which is rare. What's the event combination where `merge` is preferred to `head` in the cases where both are built? If that can't be done, duplicate tests in rare cases seem preferable to losing the merge runs.

What would be the behavior if `on` were unspecified?

From [here](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-multiple-events-with-activity-types-or-configuration), maybe we should have opt-in for branch builds and use:

```yaml
on:
  push:
    branches:
      - master
  pull_request:
```

so all pull requests are built but only default-branch pushes are built (this will need updating if we have backport branches, other tests that want to run without open PRs, unlike what we have now, I think).

Reply:
:-1: on limiting tests to branches. It's a big hindrance to anyone wanting to run tests on their own forks and branches without opening a PR. For instance if you're playing around with a few different ideas and want your commits to be CI tested you either have to edit the CI config in all of your own branches to build your other branches, or open lots of potentially pointless PRs against the parent repo.

Reply:
Yeah, I don't think we need to change the triggers. As far as the billing goes, I find this paragraph a little confusing, but the first line is "free for public repos" which I assume is equivalent to the usual statement of "always free for public repos and pay for private repos" https://docs.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions#about-billing-for-github-actions Let's burn the minutes and see if we run into any problems 😄

Reply:
Good point about branches not being up to date with master!   I usually turn on branch protection rules (which I forgot to mention) to solve that issue (under settings, screenshot below) that allows you to be confident that any tests are valid.  

![image](https://user-images.githubusercontent.com/1483922/90026537-49259580-dc6c-11ea-82e9-9d05855bdceb.png)

But feel free to do both incase there is a use case I am missing beyond branches being updated to master (I just couldn't think of any)!

Reply:
I've been tempted to suggest the latter (require branches to be up to date) but I think it imposes an unnecessary burden on external contributors, especially as we usually don't review PRs immediately. This means there are likely to be concurrent PRs, and as soon as one is merged all other PRs need to be updated. I think test with `pull_request`/`merge` (and if necessary close/open a PR to force a retest against the latest branch) is more friendly.

Reply:
@hamelsmu Just wanted to add that despite me being against your suggestions I think it's very valuable input, and would definitely be useful elsewhere. Would be nice if we could record this discussion somewhere instead of it disappearing into the ether when this PR is merged 😀.

Reply:
No problem at all!  I am by no means attached to anything and am also learning by your thoughts and comments which are very valuable.  I can see the merits of the other approaches as well which I didn't consider before!

Reply:
I also agree with @manics that the require-up-to-date rule is a contributor-busywork vs test rigor trade-off that doesn't quite pay off for our projects, but I now know it's an option, which is great, thanks @hamelsmu!

Thanks for everyone's input! This is super valuable, as there's always a variety of good ways to work with these tools. If I understood the latest comments correctly, I think we want to keep both pull_request and push triggers. Does that means there isn't more to do on this front?

It would definitely be useful to jot down some of the notes about events and what they mean and what downsides there are, e.g. summarizing from this discussion:

- pull_request builds `merge` commits, which is useful as other PRs are merged
- push events can be sufficient and may eliminate the benefit of pull_request building `merge` if branches are required to be up-to-date
- requiring branches to be up-to-date adds some work for contributors, in exchange for ensuring that tests have been run with *exactly* what will result after merge. Downside: extra work for contributors if maintainers take a while to merge PRs, as authors keep needing to update their branch while waiting for maintainers to review & merge
- push may not trigger on close/reopen PR, which is a common "poke CI" practice, especially for contributors who don't have access to the "re-run CI button"  (is this correct?)
- push & pull_request together can result in fairly redundant builds (HEAD and MERGE may differ slightly) when a PR is made from within the target repo
  - avoiding redundant builds by restricting push triggers to the default branch make it harder to test one's own forks without opening PRs (I do this a lot, including in writing this PR!)

Does that sound right?

Reply:
Yes, with the caveat that "pull_request builds merge commits, which is useful as other PRs are merged" will be a merge at the time of the build, so if another PR is later merged you need to re-trigger CI. On travis you need to close/reopen the PR to force a build of the updated "merge", it's not enough to just re-run the last job. I guess it's the same with GH actions.

Reply:
Yes, "opened", "closed" and "reopened" are all events that will trigger GHA under `pull_request` https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request